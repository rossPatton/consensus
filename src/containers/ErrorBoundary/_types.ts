export type tProps = {
  error?: any,
  isSubPage?: boolean,
  status?: tStatusUnion,
}

export type tState = tProps & {
  info: tInfoUnion | null,
};
