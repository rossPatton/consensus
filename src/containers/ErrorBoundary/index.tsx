import loglevel from 'loglevel';
import React, {ErrorInfo, PureComponent} from 'react';

import {ErrorPageComponent} from '~app/routes/ErrorPage';
import {NoMatchComponent} from '~app/routes/NoMatch';
import {UnAuthorizedComponent} from '~app/routes/UnAuthorized';

import {tProps, tState} from './_types';

export default class ErrorBoundary extends PureComponent<tProps, tState> {
  static defaultProps = {
    error: null as any,
    status: 200 as tStatusUnion,
  };

  state = {
    error: this.props.error,
    info: null as tInfoUnion,
  };

  componentDidCatch(error: Error, info: ErrorInfo) {
    this.setState({
      error,
      info,
    }, () => loglevel.error(error, info));
    // TODO log to separate service, but only show user 500 page
  }

  render() {
    const {isSubPage, status} = this.props;

    // @TODO perhaps expand the options here a bit
    // we put error boundaries on subpages as well, and instead of blowing up the page
    // we just render a small error message in place instead
    // a 204 here is not a problem really, a subpage for upcoming events just
    // might not have any, so we only care about codes over 400
    if (isSubPage && status === 500) {
      return (
        <>Something went wrong loading this section. Try refreshing?</>
      );
    }

    if (!isSubPage) {
      if (status === 500) return <ErrorPageComponent />;

      // SSR will just redirect, but sometimes on client we need to render this
      if (status === 401) return <UnAuthorizedComponent />;

      // treat 204 status as 404 on front-end
      // nothing went wrong, the route and params are correct, just no data
      if (status === 204) return <NoMatchComponent />;
    }

    return this.props.children;
  }
}
