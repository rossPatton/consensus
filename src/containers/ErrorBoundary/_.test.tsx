import React from 'react';
import render from 'react-test-renderer';

import ErrorBoundary from '.';

describe('containers/ErrorBoundary', () => {
  it('renders 200', () => {
    const component = render.create((
      <ErrorBoundary
        children=""
        status={200}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders 401', () => {
    const component = render.create((
      <ErrorBoundary
        children=""
        status={401}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders 204', () => {
    const component = render.create((
      <ErrorBoundary
        children=""
        status={204}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders 500', () => {
    const component = render.create((
      <ErrorBoundary
        children={<></>}
        status={500}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders 400', () => {
    const component = render.create((
      <ErrorBoundary
        children=""
        status={400}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders subpage', () => {
    const component = render.create((
      <ErrorBoundary
        children=""
        isSubPage
        status={500}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
