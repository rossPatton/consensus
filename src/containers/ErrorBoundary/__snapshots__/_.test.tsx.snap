// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/ErrorBoundary renders 200 1`] = `""`;

exports[`containers/ErrorBoundary renders 204 1`] = `null`;

exports[`containers/ErrorBoundary renders 400 1`] = `""`;

exports[`containers/ErrorBoundary renders 401 1`] = `null`;

exports[`containers/ErrorBoundary renders 500 1`] = `null`;

exports[`containers/ErrorBoundary renders subpage 1`] = `
<div
  className="p-2"
>
  Something went wrong loading this section. Try refreshing?
</div>
`;
